src/dune_rules/compilation_context.mli

```diff
@@ -54,8 +54,7 @@ val dir : t -> Path.Build.t
 val obj_dir : t -> Path.Build.t Obj_dir.t
 val modules : t -> Modules.With_vlib.t
 val flags : t -> Ocaml_flags.t
-val requires_link : t -> Lib.t list Resolve.Memo.t
-val requires_compile : t -> Lib.t list Resolve.Memo.t
+val requires : t -> (Lib.t * bool) list Resolve.Memo.t
```

Comment:
I think it would make sense to only change `requires_compile` to return `(Lib.t * bool) list` and keep `requires_link` returning `Lib.t list`. This will make the diff smaller and, after all, it is more correct, since for linking we don't care about whether a dependency is direct or transitive.